code/game/objects/structures/mannequin.dm

```diff
@@ -472,399 +417,246 @@
+			if(bloodsies_state)
+				var/image/bloodsies	= image('icons/effects/blood.dmi', bloodsies_state)
+				bloodsies.color		= clothToUpdate.blood_color
+				O.overlays += bloodsies
 
+		clothToUpdate.generate_accessory_overlays(O)
 
-/obj/structure/mannequin/proc/update_icon_hand(index)
-	var/obj/Overlays/hand_layer/O = obj_overlays["[MANNEQUIN_HAND_LAYER]-[index]"]
-	if(!O)
-		O = getFromPool(/obj/Overlays/hand_layer)
-		obj_overlays["[MANNEQUIN_HAND_LAYER]-[index]"] = O
-	else
-		overlays.Remove(O)
-		O.overlays.len = 0
-
-	var/obj/item/I = get_held_item_by_index(index)
+/obj/structure/mannequin/proc/update_icon_hand(var/obj/Overlays/O,var/index)
+	var/obj/item/heldItem = get_held_item_by_index(index)
 
-	if(I)
-		var/t_state = I.item_state
-		var/t_inhand_state = I.inhand_states[get_direction_by_index(index)]
+	if(heldItem)
+		var/t_state = heldItem.item_state
+		var/t_inhand_state = heldItem.inhand_states[get_direction_by_index(index)]
 		var/icon/check_dimensions = new(t_inhand_state)
 		if(!t_state)
-			t_state = I.icon_state
+			t_state = heldItem.icon_state
 
-		O.name = "[index]"
-		O.icon = t_inhand_state
-		O.icon_state = t_state
-		O.pixel_x = -1*(check_dimensions.Width() - WORLD_ICON_SIZE)/2
-		O.pixel_y = -1*(check_dimensions.Height() - WORLD_ICON_SIZE)/2
-		O.layer = O.layer
+		var/image/I  = image(t_inhand_state, t_state)
+		I.pixel_x = -1*(check_dimensions.Width() - WORLD_ICON_SIZE)/2
+		I.pixel_y = -1*(check_dimensions.Height() - WORLD_ICON_SIZE)/2
 
 		var/list/offsets = get_item_offset_by_index(index)
 
-		O.pixel_x += offsets["x"]
-		O.pixel_y += offsets["y"]
+		I.pixel_x += offsets["x"]
+		I.pixel_y += offsets["y"]
 
-		O.pixel_x += clothing_offset_x
-		O.pixel_y += clothing_offset_y
+		if(heldItem.dynamic_overlay && heldItem.dynamic_overlay["[HAND_LAYER]-[index]"])
+			var/image/dyn_overlay = heldItem.dynamic_overlay["[HAND_LAYER]-[index]"]
+			O.overlays += dyn_overlay
 
-		if(I.dynamic_overlay && I.dynamic_overlay["[MANNEQUIN_HAND_LAYER]-[index]"])
-			var/image/dyn_overlay = I.dynamic_overlay["[MANNEQUIN_HAND_LAYER]-[index]"]
-			O.overlays.Add(dyn_overlay)
+		O.overlays += I
 
-		overlays.Add(O)
+/obj/structure/mannequin/proc/get_slot_icons(var/slot)
+	switch(slot)
+		if(SLOT_MANNEQUIN_ICLOTHING)
+			return 'icons/mob/uniform.dmi'
+		if(SLOT_MANNEQUIN_FEET)
+			return 'icons/mob/feet.dmi'
+		if(SLOT_MANNEQUIN_GLOVES)
+			return 'icons/mob/hands.dmi'
+		if(SLOT_MANNEQUIN_EARS)
+			return 'icons/mob/ears.dmi'
+		if(SLOT_MANNEQUIN_OCLOTHING)
+			return 'icons/mob/suit.dmi'
+		if(SLOT_MANNEQUIN_EYES)
+			return 'icons/mob/eyes.dmi'
+		if(SLOT_MANNEQUIN_BELT)
+			return 'icons/mob/belt.dmi'
+		if(SLOT_MANNEQUIN_MASK)
+			return 'icons/mob/mask.dmi'
+		if(SLOT_MANNEQUIN_HEAD)
+			return 'icons/mob/head.dmi'
+		if(SLOT_MANNEQUIN_BACK)
+			return 'icons/mob/back.dmi'
+		if(SLOT_MANNEQUIN_ID)
+			return 'icons/mob/ids.dmi'
+		else
+			return null
 
+/obj/structure/mannequin/proc/get_primitive_icons(var/slot)
```

Comment:
This should surely inherit from the proc above somehow, as it's copy pasted from there.